include/xtensor/xnpy.hpp

```diff
@@ -0,0 +1,629 @@
+                                     (uint32_t) (header_len_le32[2] << 16) | (uint32_t) (header_len_le32[3] << 24);
+
+            if ((magic_string_length + 2 + 4 + header_length) % 16 != 0)
+            {
+                // TODO: display warning
+            }
+
+            char* buf = new char[header_length];
+            istream.read(buf, header_length);
+            std::string header(buf, header_length);
+            delete[] buf;
+
+            return header;
+        }
+
+        struct npy_file
+        {
+            npy_file() = default;
+
+            npy_file(std::vector<std::size_t>& shape, bool fortran_order,
+                     std::string typestring)
+                : m_shape(shape), m_fortran_order(fortran_order), m_typestring(typestring)
+            {
+                // Allocate memory
+                m_word_size = (std::size_t) atoi(&typestring[2]);
+                m_n_bytes = compute_size(shape) * m_word_size;
+                m_buffer = new char[m_n_bytes];
+            }
+
+            // TODO REMOVE COPY!
+            template <class T, layout_type L = layout_type::dynamic>
+            xarray<T, L> cast(bool check_type = true)
+            {
+                T* ptr = reinterpret_cast<T*>(&m_buffer[0]);
+                xbuffer_adaptor<T, acquire_ownership> buffer_adapter(ptr, compute_size(m_shape));
+                std::vector<std::size_t> strides(m_shape.size());
+
+                // check if the typestring matches the given one
+                if (check_type && m_typestring != detail::build_typestring<T>())
+                {
+                    throw std::runtime_error("formatting error: template format not matching"s + m_typestring + 
+                                             " vs "s + detail::build_typestring<T>());
+                }
+
+                if ((L == layout_type::column_major && !m_fortran_order) ||
+                    (L == layout_type::row_major && m_fortran_order))
+                {
+                    throw std::runtime_error("fortran order and xarray layout type do not match up.");
+                }
+
+                compute_strides(m_shape,
+                                m_fortran_order ? layout_type::column_major : layout_type::row_major,
+                                strides);
+                std::vector<std::size_t> shape(m_shape);
+                return xarray_container<decltype(buffer_adapter), L, std::vector<std::size_t>>(
+                    std::move(buffer_adapter), std::move(shape), std::move(strides));
+            }
+
+            char* ptr()
+            {
+                return m_buffer;
+            }
+
+            std::size_t n_bytes()
+            {
+                return m_n_bytes;
+            }
+
+            std::vector<std::size_t> m_shape;
+            bool m_fortran_order;
+            size_t m_word_size;
+            size_t m_n_bytes;
+            std::string m_typestring;
+            char* m_buffer;
```

Comment:
As discussed there is a memory leak here. 

I see two options: 
a) adding a boolean (pointer_moved, or has_been_cast) that indicates wether this object has been cast, and the pointer is owned by the xadaptor (which will take care of the destruction).
b) using a smart pointer (maybe `unique`, or `shared`) and hope that the xadaptor can handle them ;)

What do you think @SylvainCorlay?

Reply:
Maybe the npy file could have two actual C++ casting operators

- one that return an `xarray_container<owning_buffer_adapter, L, std::vector<std::size_t>> &&` which can only be called when the npy file is an rvalue. It takes ownership of the pointer and prevents the npy file destructor to actually delete the buffer with a `moved_from` flag or setting the pointer in the class to `nullptr`.

- one to `xarray_adaptor<nonowning_buffer_adapter, L, std::vector<std::size_t>>` which can only be called when the npy file is an lvalue. It does not take ownership of the pointer. When assigning it to a container, the generalized copy constructor is called and a copy occurs.